backend/app/Models/BatchStep.php

```
@@ -198,6 +205,31 @@ public function pendingRequiredChecklistLabels(): \Illuminate\Support\Collection
+
+        $required = TemplateStepOutput::where('process_template_id', $templateId)
+            ->where('is_required', true)
+            ->whereHas('templateStep', fn ($q) => $q->where('step_number', $this->step_number))
+            ->pluck('label', 'id');
+        if ($required->isEmpty()) {
+            return collect();
+        }
+
+        $recorded = $this->outputValues()->pluck('output_id')->all();
+
+        return $required->reject(fn ($label, $id) => in_array($id, $recorded, true))->values();
```

PHPMD (2.15.0)

[warning] Avoid unused parameters such as '$label'. (undefined)

(UnusedFormalParameter)